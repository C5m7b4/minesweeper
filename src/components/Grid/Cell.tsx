import React, { FC } from 'react';
import styled from '@emotion/styled';

import { Cell as CellType, CellState, Coords } from '@/helpers/Field';

export interface CellProps {
  /**
   * Cell Status based on Celltype
   */
  children: CellType;
  /**
   * Cell Coordinates
   */
  coords: Coords;
  /**
   * onClick cell handler
   */
  onClick: (coords: Coords) => void;
  /**
   * onContext Menu
   */
  onContextMenu: (coords: Coords) => void;
}

export const Cell: FC<CellProps> = ({ children, coords, ...rest }) => {
  const isActiveCell = [
    CellState.hidden,
    CellState.mark,
    CellState.weakMark,
  ].includes(children);

  const onClick = () => {
    if (isActiveCell) {
      rest.onClick(coords);
    }
  };

  const onContextMenu = (elem: React.MouseEvent<HTMLElement>) => {
    /**
     * Prevent context menu by default
     */
    elem.preventDefault();
    if (isActiveCell) {
      rest.onContextMenu(coords);
    }
  };

  const props = {
    onClick,
    onContextMenu,
  };

  return <ComponentsMap {...props}>{children}</ComponentsMap>;
};

interface ComponentsMapProps {
  children: CellType;
  onClick: (elem: React.MouseEvent<HTMLElement>) => void;
  onContextMenu: (elem: React.MouseEvent<HTMLElement>) => void;
}

const ComponentsMap: FC<ComponentsMapProps> = ({ children, ...rest }) => {
  switch (children) {
    case CellState.empty:
      return <ReveledFrame {...rest} />;
    case CellState.mark:
      return (
        <ClosedFrame {...rest}>
          <Flag />
        </ClosedFrame>
      );
    case CellState.weakMark:
      return (
        <ClosedFrame {...rest}>
          <WeakFlag />
        </ClosedFrame>
      );
    case CellState.bomb:
      return (
        <BombFrame {...rest}>
          <Bomb />
        </BombFrame>
      );
    case CellState.hidden:
      return <ClosedFrame {...rest} />;
    default:
      return <ReveledFrame {...rest}>{children}</ReveledFrame>;
  }
};

const transparent = 'rgba(0,0,0,0)';

const colors: { [key in CellType]: string } = {
  0: transparent,
  1: '#2a48ec',
  2: '#2bb13d',
  3: '#ec6561',
  4: '#233db7',
  5: '#a6070f',
  6: '#e400af',
  7: '#906a02',
  8: '#fa0707',
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
};

const ClosedFrame = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vw;
  height: 1.8vw;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;
  &:hover {
    filter: brightness(1.1);
  }
`;

const ReveledFrame = styled(ClosedFrame)`
  border-color: #dddddd;
  cursor: default;
  color: ${({ children }) => colors[children as CellType] ?? transparent};
  &:hover {
    filter: brightness(1);
  }
`;

const Bomb = styled.div`
  border-radius: 50%;
  width: 1vw;
  height: 1vw;
  background-color: #333;
`;

const BombFrame = styled(ReveledFrame)`
  background-color: #ec433c;
`;

const Flag = styled.div`
  width: 0px;
  height: 0px;
  color: ${transparent};
  border-top: 2vh solid transparent;
  border-bottom: 2vh solid transparent;
  border-left: 2vh solid #ec433c;
`;

const WeakFlag = styled(Flag)`
  border-left: 2vh solid #f19996;
`;
