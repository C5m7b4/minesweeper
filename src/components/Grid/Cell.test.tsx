import React from 'react';
import { render, screen, fireEvent, createEvent } from '@testing-library/react';

import { CellState, Coords } from '@/helpers/Field';
import { Cell, checkCellIsActive } from './Cell';

describe('Cell Component Check', () => {
  const coords: Coords = [1, 1];
  for (let cell = CellState.empty; cell <= CellState.weakMark; cell++) {
    it('Check prevent default context menu for every type of cell', () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<Cell {...props}>{cell}</Cell>);
      const cellCmp = screen.getByTestId(`${cell}_${coords}`);
      const contextMenuEvent = createEvent.contextMenu(cellCmp);
      fireEvent(cellCmp, contextMenuEvent);
      expect(contextMenuEvent.defaultPrevented).toBe(true);
    });
    it('onClick and onContextMenu handle should be called for active cells', () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<Cell {...props}>{cell}</Cell>);
      const cellCmp = screen.getByTestId(`${cell}_${coords}`);

      fireEvent.click(cellCmp);
      fireEvent.contextMenu(cellCmp);
      if (checkCellIsActive(cell)) {
        expect(props.onClick).toBeCalled();
        expect(props.onContextMenu).toBeCalled();
      } else {
        expect(props.onClick).not.toBeCalled();
        expect(props.onContextMenu).not.toBeCalled();
      }
    });
  }
});
