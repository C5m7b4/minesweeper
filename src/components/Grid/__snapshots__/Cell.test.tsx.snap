// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Component Check Closed Frame should render correctly 1`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 2`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 3`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 4`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 5`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 6`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 7`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 8`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 9`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 10`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 11`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 12`] = `[Function]`;

exports[`Cell Component Check Closed Frame should render correctly 13`] = `[Function]`;
